lib/segment/src/index/hnsw_index/hnsw.rs

```diff
@@ -349,13 +370,23 @@ impl HNSWIndex {
         if build_main_graph {
             let timer = std::time::Instant::now();
 
-            let mut ids_iterator = id_tracker_ref.iter_ids_excluding(deleted_bitslice);
+            let mut ids = Vec::with_capacity(total_vector_count);
+            if let Some(old_index) = &old_index {
+                for vector_id in id_tracker_ref.iter_ids_excluding(deleted_bitslice) {
+                    if let Some(links) = old_index.get_links(vector_id) {
+                        graph_layers_builder.add_new_point(vector_id, links);
+                    } else {
+                        ids.push(vector_id);
+                    }
+                }
+            } else {
+                ids.extend(id_tracker_ref.iter_ids_excluding(deleted_bitslice));
+            }
 
-            let first_few_ids: Vec<_> = ids_iterator
-                .by_ref()
-                .take(SINGLE_THREADED_HNSW_BUILD_THRESHOLD)
-                .collect();
-            let ids: Vec<_> = ids_iterator.collect();
+            let first_few_ids = ids.split_off(
```

Comment:
`first_few_ids` is only relevant if the graph is empty. Potentially we can skip this step if we have already added enough points with `add_new_point`

Reply:
Updated.